icefwd.1
.\" Copyright (C) 2014 Jyzee <jyzee.git(at)gmail(dot)com>
.\" Copyright (C) 2006 Moritz Grimm <gtgbr@gmx.net>
.\"
.\" The following legalese applies to this manual only:
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF MIND, USE, DATA OR PROFITS, WHETHER IN
.\" AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT
.\" OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd October 19, 2016
.Dt ICEFWD 1
.Os
.Sh NAME
.Nm icefwd
.Nd "pipe an Ogg/WebM stream from stdin to an Icecast server"
.Sh SYNOPSIS
.Nm
.Bk -words
.Op Fl hpv
.Op Fl f Ar format
.Op Fl d Ar description
.Op Fl m Ar metadata file
.Op Fl g Ar genre
.Op Fl n Ar name
.Op Fl u Ar URL
.Op Fl T Ar TLS mode
.Ar address
.Ar port
.Ar password
.Ar mountpoint
.Ekq
.Sh DESCRIPTION
The
.Nm
utility acts as a minimal source client for Icecast servers.
It reads an Ogg/WebM stream from stdin and forwards it to a server specified on the
command line.
.Pp
The
.Ar address ,
.Ar port ,
.Ar password
and
.Ar mountpoint
arguments are mandatory and have to be given in that order.
.Ss Optional command line parameters
.Bl -tag -width Ds
.It Fl h
Print a short usage summary.
.It Fl v
Print version.
.It Fl n Ar name
Set the stream
.It Fl f Ar format
Set the stream format. Available formats are ogg or webm, default is ogg.
.It Fl d Ar description
Set the stream description.
.It Fl m Ar metadata file
Set a file holding stream meta data.
.It Fl g Ar genre
Set the stream genre.
.Pq i.e. radio
name.
.It Fl p
Allow the stream to be made public in stream directories.
.It Fl u Ar URL
Set the stream URL.
.It Fl T Ar TLS mode
This sets the TLS (encryption) mode. Valid values are "disabled" (no encryption), "auto" (any encryption if available), "auto_no_plain" (any encryption), "rfc2818" (encryption based on RFC2818) and "rfc2817" (encryption based on RFC2817). For most users the default mode "auto" should work just fine.
.El
.Pp
.Nm
reacts to the hangup signal, SIGHUP, by printing the amount of bytes read from
stdin to stdout at the next given opportunity.
How fast it reacts depends on the stream's current bitrate.
.Sh EXAMPLES
To forward a public Ogg stream with the name
.Dq Test radio
to an Icecast server, one would pipe that stream to an
.Nm
command line such as
.Pp
.Bd -literal -offset 4n
icefwd -p -n \&"Test radio\&" radio.example.com 8000 password /stream.ogg
.Ed
.Pp
Encode an Ogg Vorbis stream, dump it to disk and stream it at the same time:
.Pp
.Bd -literal -offset 4n
oggenc - < input.wav | tee streamdump.ogg \e
    | icefwd radio.example.com 8000 password /stream.ogg
.Ed
.Pp
Stream WebM to Icecast from a pipe created with mkfifo:
.Pp
.Bd -literal -offset 4n
icefwd -f webm radio.example.com 8000 password /stream.webm < /tmp/streampipe
.Ed
.Sh AUTHORS
Programming by
.Bl -item -offset indent -compact
.It
Jyzee <jyzee.git(at)gmail(dot)com>
.It
J^ <j@v2v.cc>
.It
rafael2k <rafael@riseup.net>
.It
Moritz Grimm <gtgbr@gmx.net>
.It
Philipp Schafft <lion@lion.leolix.org>
.El
.Pp
This manual was contributed by
.An Moritz Grimm and updated by Philipp Schafft.
.An Edited by Jyzee for icefwd.
.Sh CAVEATS
Since the password to the Icecast server is given in clear text on the command
line, other
.Pq local
users will probably be able to see it very easily.


ChangeLog
0.4.0
  * Added support for TLS.
  * Fix man page.
0.3.1
  * Use SHOUT_FORMAT_OGG instead of SHOUT_FORMAT_VORBIS (deprecated)
  * Makefile: fix build on Ubuntu (and probably others)
0.3.0
  * Add WebM support
  * Non-required dependencies removed
  * Man page updated for icefwd
  * Forked from Oggfwd
0.2
  * Makefile: rewrite; more packager-friendly
  * man page
  * use getopt() for advanced settings
  * remove "dump file" feature, it is not implemented in libshout. explain
    tee(1) usage instead
  * signal handling to print "total bytes read" stats
  * write all errors to stderr, regular output to stdout. Redirecting output
    to /dev/null makes oggfwd silent, except when errors occur
  * lint + -Wall clean
0.1
  * add advanced features by rafael2k <rafael@riseup.net>
  * initial version


icefwd.c
/*
 * icefwd -- Forward an Ogg/WebM stream from stdin to an Icecast server
 *    A useful demonstration of the libshout API
 *
 * This program is distributed under the GNU General Public License, version 2.
 * A copy of this license is included with this source.
 *
 * This program is provided "as-is", with no explicit or implied warranties of
 * any kind.
 *
 * Copyright (C) 2003-2006, J <j@v2v.cc>,
 *                          rafael2k <rafael(at)riseup(dot)net>,
 *                          Moritz Grimm <gtgbr@gmx.net>
 * Copyright (C) 2014-2016  Jyzee <jyzee.git(at)gmail(dot)com>
 * Copyright (C) 2015       Philipp Schafft <lion@lion.leolix.org>
 *
 * thanx to rhatto <rhatto (AT) riseup (DOT) net> and others at submidialogia :-P
 */

#include <sys/types.h>
#include <sys/param.h>
#include <errno.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <fcntl.h>

#ifndef NO_UNISTD_H
#include <unistd.h>
#endif

#include <shout/shout.h>

extern char *__progname;
extern char *optarg;
extern int optind;
extern int errno;

volatile sig_atomic_t print_total = 0;
volatile sig_atomic_t quit = 0;

#define METABUFSIZE 4096
char *metafilename;
shout_t *shout;

#define BUFFERSIZE 4096

#if defined(__dead)
__dead void
#else
void
#endif /* __dead */
usage(void)
{
	printf("Usage: %s "
	       "[-hpv] "
	       "[-f format] "
	       "[-d description] "
	       "[-m metadata file] "
	       "[-g genre] "
	       "[-n name] "
	       "[-u URL] "
#ifdef SHOUT_TLS
	       "[-T {disabled|auto|auto_no_plain|rfc2818|rfc2817}] "
#endif
	       "\n              address port password mountpoint\n",
	       __progname);
	exit(1);
}

#if defined(__dead)
__dead void
#else
void
#endif /* __dead */
version(void)
{
	printf("IceFWD version 0.4.0\n");
	exit(1);
}

void load_metadata() {
	int i, fh, r;
	char buf[METABUFSIZE], *key, *val;
	enum {
		state_comment,
		state_key,
		state_value,
		state_unknown
	} state;

	bzero(buf, METABUFSIZE);

	if (!metafilename) {
		fprintf(stderr, "Please use the -m argument to set the meta file name!\n");
		return;
	}

	fh = open(metafilename, O_RDONLY);

	if (-1 == fh) {
		fprintf(stderr, "Error while opening meta file \"%s\": %s\n", metafilename, strerror(errno));
		return;
	}

	r = read(fh, &buf, METABUFSIZE);
	if (-1 == r) {
		fprintf(stderr, "Error while reading meta file \"%s\": %s\n", metafilename, strerror(errno));
		close(fh);
		return;
	}

	state = state_unknown;
	key = val = NULL;
	i = 0;

	while (i < METABUFSIZE) {
		switch (buf[i]) {
			case 0:
				/* we're done */
				i = METABUFSIZE;
				break;
			case '\r':
			case '\n':
				if (state_value == state) {
					buf[i] = 0;

					if (key && val) {
						if (strcmp("name", key) == 0) {
							shout_set_name(shout, val);
						} else if (strcmp("genre", key) == 0) {
							shout_set_genre(shout, val);
						} else if (strcmp("description", key) == 0) {
							shout_set_description(shout, val);
						} else if (strcmp("url", key) == 0) {
							shout_set_url(shout, val);
						}
					}
				}
				state = state_unknown;
				key = NULL;
				val = NULL;
				break;
			case '=':
				if (state_key == state) {
					buf[i] = 0;
					state = state_value;
					val = &buf[i+1];
				}
				break;
			case '#':
				if (state_unknown == state) {
					state = state_comment;
				}
				break;
			default:
				if (state_unknown == state) {
					state = state_key;
					key = &buf[i];
				}
		}

		i++;
	}

	close(fh);
}

void sig_handler(int sig) {
	switch (sig) {
		case SIGHUP:
			print_total = 1;
			break;
		case SIGTERM:
		case SIGINT:
			quit = 1;
			break;
		case SIGUSR1:
			load_metadata();
			break;
	}
}

void set_argument_string(char **param, char *opt, char optname) {
	size_t siz;

	if (*param) {
		fprintf(stderr, "%s: Parameter -%c given multiple times\n", __progname, optname);
		usage();
	}

	siz = strlen(opt) + 1;
	if (siz >= MAXPATHLEN) {
		fprintf(stderr, "%s: Argument for parameter -%c too long\n", __progname, optname);
		exit(1);
	}

	if ((*param = malloc(siz)) == NULL) {
		fprintf(stderr, "%s: %s\n", __progname, strerror(errno));
		exit(1);
	}

	snprintf(*param, siz, "%s", opt);
}

#ifdef SHOUT_TLS
void set_tls_mode(int *tls_mode, char *opt, char optname) {
	if (0==strcasecmp("DISABLED", opt)) {
		*tls_mode = SHOUT_TLS_DISABLED;
	} else if (0==strcasecmp("AUTO", opt)) {
		*tls_mode = SHOUT_TLS_AUTO;
	} else if (0==strcasecmp("AUTO_NO_PLAIN", opt)) {
		*tls_mode = SHOUT_TLS_AUTO_NO_PLAIN;
	} else if (0==strcasecmp("RFC2818", opt)) {
		*tls_mode = SHOUT_TLS_RFC2818;
	} else if (0==strcasecmp("RFC2817", opt)) {
		*tls_mode = SHOUT_TLS_RFC2817;
	} else {
		fprintf(stderr, "%s: Invalid value for -%c.\n", __progname, optname);
		usage();
		exit(1);
	}
}
#endif

int main(int argc, char **argv) {
	unsigned char buff[BUFFERSIZE];
	int ret, ch;
	unsigned int pFlag;
	char *format, *description, *genre, *name, *url;
	size_t bytes_read = 0;
	unsigned short port;
	unsigned long long total;
#ifdef SHOUT_TLS
	int tls_mode = SHOUT_TLS_AUTO;
#endif

	pFlag = 0;
	format = description = genre = name = url = metafilename = NULL;
	while ((ch = getopt(argc, argv, "vd:g:hn:m:pu:f:T:")) != -1) {
		switch (ch) {
			case 'v':
				version();
				break;
			case 'd':
				set_argument_string(&description, optarg, 'D');
				break;
			case 'g':
				set_argument_string(&genre, optarg, 'g');
				break;
			case 'n':
				set_argument_string(&name, optarg, 'n');
				break;
			case 'm':
				set_argument_string(&metafilename, optarg, 'm');
				break;
			case 'p':
				pFlag = 1;
				break;
			case 'u':
				set_argument_string(&url, optarg, 'u');
				break;
			case 'f':
				set_argument_string(&format, optarg, 'f');
				break;
			case 'T':
#ifdef SHOUT_TLS
				set_tls_mode(&tls_mode, optarg, 'T');
				break;
#endif
			case 'h':
			default:
				usage();
		}
	}
	argc -= optind;
	argv += optind;

	if (argc == 0)
		usage();

	if (argc != 4) {
		fprintf(stderr, "%s: Wrong number of arguments\n", __progname);
		usage();
	}

	if ((shout = shout_new()) == NULL) {
		fprintf(stderr, "%s: Could not allocate shout_t\n", __progname);
		return 1;
	}

#ifdef SHOUT_TLS
	if (shout_set_tls(shout, tls_mode) != SHOUTERR_SUCCESS) {
		fprintf(stderr, "%s: Error setting TLS mode: %s\n", __progname, shout_get_error(shout));
		return 1;
	}
#endif

	if (shout_set_host(shout, argv[0]) != SHOUTERR_SUCCESS) {
		fprintf(stderr, "%s: Error setting hostname: %s\n", __progname, shout_get_error(shout));
		return 1;
	}

	if (sscanf(argv[1], "%hu", &port) != 1) {
		fprintf(stderr, "Invalid port `%s'\n", argv[1]);
		usage();
	}

	if (shout_set_port(shout, port) != SHOUTERR_SUCCESS) {
		fprintf(stderr, "%s: Error setting port: %s\n", __progname, shout_get_error(shout));
		return 1;
	}

	if (shout_set_password(shout, argv[2]) != SHOUTERR_SUCCESS) {
		fprintf(stderr, "%s: Error setting password: %s\n", __progname, shout_get_error(shout));
		return 1;
	}

	if (shout_set_mount(shout, argv[3]) != SHOUTERR_SUCCESS) {
		fprintf(stderr, "%s: Error setting mount: %s\n", __progname, shout_get_error(shout));
		return 1;
	}

	if (format && strncmp(format, "webm", 4) == 0) {
		shout_set_format(shout, SHOUT_FORMAT_WEBM);
	} else {
		shout_set_format(shout, SHOUT_FORMAT_OGG);
	}

	shout_set_public(shout, pFlag);

	if (metafilename)
		load_metadata();

	if (description)
		shout_set_description(shout, description);

	if (genre)
		shout_set_genre(shout, genre);

	if (name)
		shout_set_name(shout, name);

	if (url)
		shout_set_url(shout, url);

	signal(SIGUSR1, sig_handler);

	//wait for data before opening connection to server
	bytes_read = fread(buff, 1, sizeof(buff), stdin);

	if (shout_open(shout) == SHOUTERR_SUCCESS) {
		printf("%s: Connected to server\n", __progname);
		total = 0;

		signal(SIGHUP, sig_handler);
		signal(SIGTERM, sig_handler);
		signal(SIGINT, sig_handler);

		while (quit == 0) {
			total += bytes_read;

			if (bytes_read > 0) {
				ret = shout_send(shout, buff, bytes_read);
				if (ret != SHOUTERR_SUCCESS) {
					printf("%s: Send error: %s\n", __progname, shout_get_error(shout));
					quit = 1;
				}
			} else {
				quit = 1;
			}

			if (quit) {
				printf("%s: Quitting ...\n", __progname);
				print_total = 1;
			}

			if (print_total) {
				printf("%s: Total bytes read: %llu\n", __progname, total);
				print_total = 0;
			}

			shout_sync(shout);

			bytes_read = fread(buff, 1, sizeof(buff), stdin);
			if (bytes_read != sizeof(buff) && feof(stdin))
				quit = 1;
		}
	} else {
		fprintf(stderr, "%s: Error connecting: %s\n", __progname, shout_get_error(shout));
		return 1;
	}

	shout_close(shout);

	return 0;
}
